Update the existing salt.md, add the start of a Salt README tree

zmerlynn opened the pull request:
Starts to fix #6070

zmerlynn committed:
Update the existing salt.md, add the start of a Salt README tree

Starts to fix #6070

mbforbes commented:
Thanks Zach. Reviewing now.

mbforbes commented on cluster/saltbase/README.md:
```diff
@@ -0,0 +1,12 @@
+This SaltStack configuration currently applies to default
+configurations for Debian-on-GCE, Fedora-on-Vagrant, Ubuntu-on-AWS and
+Ubuntu-on-Azure. (That doesn't mean it can't be made to apply to an
+arbitrary configuration, but those are only in-tree supported OS/IaaS
```
add "the," remove "supported:" _"those are **the** only in-tree OS/IaaS"_

mbforbes commented on cluster/saltbase/README.md:
```diff
@@ -0,0 +1,12 @@
+arbitrary configuration, but those are only in-tree supported OS/IaaS
+combinations supported today.) As you peruse the configuration, these
+are shorthanded as `gce`, `vagrant`, `aws`, `azure` in `grains.cloud`;
+the documentation in this tree uses this same shorthand, for convenience.
```
comma not needed

mbforbes commented on cluster/saltbase/README.md:
```diff
@@ -0,0 +1,12 @@
+configurations for Debian-on-GCE, Fedora-on-Vagrant, Ubuntu-on-AWS and
+Ubuntu-on-Azure. (That doesn't mean it can't be made to apply to an
+arbitrary configuration, but those are only in-tree supported OS/IaaS
+combinations supported today.) As you peruse the configuration, these
```
"As you peruse" comes as a surprise&mdash;what am I perusing? I think a note in the first paragraph that you can dive into sub-directories (especially `salt/`) for lots more information would help people know where to look for further documentation. (I worry that a project newbie wouldn't know specifically to look in `cluster/saltbase/salt/` to find the detailed docs you just wrote!)

zmerlynn replied:
I actually originally had each directory described, then I realized they belonged in readmes in the directory below. Not sure how to make it the most visible.

mbforbes replied:
A tiny brief overview, like a table of contents, would at least let people know there's more to come and where to loook

zmerlynn replied:
I added a little tiny thing, but there's a silly tension here between duplicating the exact same content that's already on the screen, and not.

mbforbes replied:
I see what you mean. I made the above comment before knowing much about salt. Now that I've read some amount of salt documentation, I see that `salt/`, `pillar/`, and `reactor/` are the three main directory trees of salt, so perusing into those specifics is obvious if you know what you're looking for. If you don't know about salt, they seem like completely arbitrary names, and one is immediately motivated to move a directory up and not keep looking.

I guess it depends who you're writing this documentation for. If it's people that know about salt, you needn't say anything more. If it's for salt newbies, some motivation for looking further is needed.

Anyway, I think the links you've added give a nice balance, so this looks good. Just wanted to provide a bit of context that may seem obvious for someone who is knee deep in salt.

mbforbes commented on cluster/saltbase/pillar/README.md:
```diff
@@ -0,0 +1,4 @@
+The
+[SaltStack pillar](http://docs.saltstack.com/en/latest/topics/pillar/)
+data is partially statically dervied from the contents of this
+directory, but largely generated at cluster inception today.
```
remove "today"

mbforbes replied:
- how is it generated at cluster inception?
- where can we go to find details on what goes in it?

zmerlynn replied:
Ah, right. This is when it became 10pm and I ran out of steam. :)

mbforbes replied:
Nice, this looks way better.

zmerlynn commented:
PTAL. I certainly have more I can add, but I believe I addressed your first round (and had time-boxed myself, so am time-boxing myself by merely addressing your comments. :))

mbforbes commented:
I'll check it out soon. I'm rounding out a shallow dive into salt's docs itself. I suspect after I'll have a handful of nits on the remaining files.

I'll post here when I'm looking again. Thanks for your patience! (It's been super helpful to actually look at salt docs... I now have _some_ idea what's happening and what things mean.)

mbforbes commented on cluster/saltbase/README.md:
```diff
@@ -0,0 +1,17 @@
+See more:
+* [pillar](pillar/)
+* [reactor](reactor/)
+* [salt](salt/)
```
reviewed

mbforbes commented on cluster/saltbase/install.sh:
```diff
@@ -15,7 +15,7 @@
+# argument that is a tarball with the pre-compiled kubernetes server binaries.
 
 set -o errexit
 set -o nounset
```
reviewed

mbforbes commented on cluster/saltbase/pillar/README.md:
```diff
@@ -0,0 +1,19 @@
+
+## Future work
+
+Document the current pillars across providers
```
reviewed

mbforbes commented on cluster/saltbase/reactor/README.md:
```diff
@@ -0,0 +1,3 @@
+[SaltStack reactor](http://docs.saltstack.com/en/latest/topics/reactor/) files, largely defining reactions to new nodes.
+
+**Ignored for GCE, which runs standalone on each machine**
```
reviewed

mbforbes commented on cluster/saltbase/salt/README.md:
```diff
@@ -0,0 +1,31 @@
+Key: M = Config applies to master, n = config applies to nodes
+
+Config                                              | GCE   | Vagrant | AWS | Azure
+----------------------------------------------------|-------|---------|-----|------
```
Cold you please add a section for `_states/`, or a brief desc. as to how it's included?

mbforbes commented on cluster/saltbase/salt/README.md:
```diff
@@ -0,0 +1,31 @@
+[debian-auto-upgrades](debian-auto-upgrades/)       | M n   | M n     | M n | M n
+[docker](docker/)                                   | M n   | M n     | M n |   n
+[etcd](etcd/)                                       | M     | M       | M   | M
+[fluentd-es](fluentd-gcp/) (pillar conditional)     | M n   | M n     | M n | M n
```
wrong link target

mbforbes commented on cluster/saltbase/salt/README.md:
```diff
@@ -0,0 +1,31 @@
+[fluentd-es](fluentd-gcp/) (pillar conditional)     | M n   | M n     | M n | M n
+[fluentd-gcp](fluentd-gcp/) (pillar conditional)    | M n   | M n     | M n | M n
+[generate-cert](generate-cert/)                     | M     | M       | M   | M
+[kube-proxy](kube-proxy/)                           |   n   |   n     |   n |   n
```
Can you please reorder to match the ordering in the file list above? It goes `-addons`, then `-apiserver`, etc.

mbforbes commented on cluster/saltbase/salt/README.md:
```diff
@@ -0,0 +1,31 @@
+[kube-addons](kube-addons/)                         | M     | M       | M   | M
+[kube-apiserver](kube-apiserver/)                   | M     | M       | M   | M
+[kube-controller-manager](kube-controller-manager/) | M     | M       | M   | M
+[kube-client-tools](kube-client-tools/)             | M     | M       | M   | M
```
`kube-client-tools` doesn't exist; please remove. Perhaps this is since I took a while to review this :-)

mbforbes replied:
Oops, now I see, it's just a raw file. Maybe update link and name and put at end (to match github file order)?

mbforbes commented on cluster/saltbase/salt/README.md:
```diff
@@ -0,0 +1,31 @@
+reading Jinja templates, the following tables break down what
+configurations run on what providers:
+
+Key: M = Config applies to master, n = config applies to nodes
```
Is it worth a little note why this is sometimes unintuitive? I.e. `kubelet` runs on most masters?

zmerlynn replied:
I can leave that for a follow-on.

mbforbes commented on cluster/saltbase/salt/README.md:
```diff
@@ -0,0 +1,31 @@
+[monit](monit/)                                     | M n   | M n     | M n | M n
+[nginx](nginx/)                                     | M     | M       | M   | M
+[openvpn](openvpn/)                                 |       |         |     | M
+[openvpn-client](openvpn-client/)                   |       |         |     |   n
```
nit: giuthub has `-client` first. Whether or not it's right, it'd be nice to fully match the ordering.

mbforbes commented on cluster/saltbase/salt/README.md:
```diff
@@ -0,0 +1,31 @@
+[nginx](nginx/)                                     | M     | M       | M   | M
+[openvpn](openvpn/)                                 |       |         |     | M
+[openvpn-client](openvpn-client/)                   |       |         |     |   n
+[sdn](sdn/) (config not active except on Vagrant)   |   n   | M n     |   n |
```
nit: the parentheses could be more succinct: "Vagrant only"

mbforbes commented on docs/salt.md:
```diff
@@ -6,11 +6,11 @@ The Salt scripts are shared across multiple hosting providers, so it's important
 ## Salt cluster setup
 
-The **salt-master** service runs on the kubernetes-master node.
+The **salt-master** service runs on the kubernetes-master node (except on default GCE setup, see below).
```
here's a cool thing: you could have both this and the next parentheses that you added just say "(except on GCE)" and have that text link to `#standalone-salt-configuration-on-gce`, which points to the paragraph you're referencing.

mbforbes commented on docs/salt.md:
```diff
@@ -20,8 +20,16 @@ The salt-master is contacted by each salt-minion and depending upon the machine
+
+On GCE, the master and nodes are all configured as [standalone minions](http://docs.saltstack.com/en/latest/topics/tutorials/standalone_minion.html). The configuration for each VM is derived from the VM's [instance metadata](https://cloud.google.com/compute/docs/metadata) and then stored in Salt grains (`/etc/salt/minion.d/grains.conf`) and pillars (`/srv/salt-overlay/pillar/cluster-params.sls`)  ...
+
+All remaining sections that refer to master/minion setups should be ignored for GCE. One fallout of the GCE setup is that the Salt mine doesn't exist - there is no sharing of configuration amongst nodes. This has ramifications later.
```
"This has ramifications later," aside from being a little ominous (which I like), isn't terribly helpful here. I would say we should either be explicit about the ramifications here, link to where the reader can find more information, or just not say this.

mbforbes commented on docs/salt.md:
```diff
@@ -20,8 +20,16 @@ The salt-master is contacted by each salt-minion and depending upon the machine
+
 ## Salt security
 
+(Not applicable on default GCE setup.)
```
Fun optional nit: I think this would look great in italics. Feel free to disagree.

mbforbes commented on docs/salt.md:
```diff
@@ -55,6 +64,8 @@ Key | Value
 `cloud` | (Optional) Which IaaS platform is used to host kubernetes, *gce*, *azure*, *aws*, *vagrant*
 `etcd_servers` | (Optional) Comma-delimited list of IP addresses the kube-apiserver and kubelet use to reach etcd.  Uses the IP of the first machine in the kubernetes_master role.
 `hostnamef` | (Optional) The full host name of the machine, i.e. hostname -f
+`etcd_servers` | (Optional) Comma-delimited list of IP addresses the kube-apiserver and kubelet use to reach etcd.  Uses the IP of the first machine in the kubernetes_master role, or 127.0.0.1 on GCE.
```
This looks like a merging bug—you've duplicated these two keys.

mbforbes commented:
This is looking great, Zach. Thanks again for doing this. Please ping after addressing my comments; it should be ready then.

zmerlynn commented:
PTAL. I think I addressed everything.

mbforbes commented:
Most things addressed&mdash;let's get this in. Thanks Zach.
